### PR TITLE
Add print-rate as optional argument

### PR DESCRIPTION
**Pull request by zombie-einstein**
This lets the print-rate be passed through as an argument, I've noticed in some cases where the scan/loop iterations are not that fast, the 1/20th can be a bit slow. It defaults back to the previous behaviour. What'd you think @jeremiecoullon ?

FYI I need to add tests and update the README if we want to add.

### CODE REVIEW

**Comment by jeremiecoullon** [comment]
Great idea! 
Maybe add a bit to the README saying that this feature exists. 

re-tests: I don't mind, it doesn't seem essential but if you want to add some then why not.

**Comment by zombie-einstein** [comment]
> Great idea! Maybe add a bit to the README saying that this feature exists.
> 
> re-tests: I don't mind, it doesn't seem essential but if you want to add some then why not.

Cool yup added to README, and just test it actually runs if  you could grab a look.

File: README.md [hunk 1]
```diff
@@ -43,6 +43,29 @@ def step(i, val):
 last_number = lax.fori_loop(0, n, step, 0)
 ```
 
+### Print Rate
+
+By default, the progress bar is updated 20 times over the course of the scan/loop
+(for performance purposes, see [below](#why-jax-tqdm)). This
+update rate can be manually controlled with the `print_rate` keyword argument. For
+example
```
Comment (jeremiecoullon) [inline README.md:51]: nitpicking: a colon might be nice here, namely: `for example:`

File: README.md [hunk 2]
```diff
@@ -43,6 +43,29 @@ def step(i, val):
...
+n = 10_000
+
+@scan_tqdm(n, print_rate=2)
+def step(carry, x):
+    return carry + 1, carry + 1
+
+last_number, all_numbers = lax.scan(step, 0, jnp.arange(n))
+```
+
+will update every-other steps.
```
Comment (jeremiecoullon) [inline README.md:67]: no need for a hyphen here; saying `every one step` is good

File: README.md [hunk 3]
```diff
@@ -43,6 +43,29 @@ def step(i, val):
...
+
 ## Why JAX-tqdm?
 
 JAX functions are [purely functional](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#pure-functions), so side effects such as printing progress when running scans and loops are not allowed. However, the [host_callback module](https://jax.readthedocs.io/en/latest/jax.experimental.host_callback.html) has primitives for calling Python functions on the host from JAX code. This can be used to update a Python tqdm progress bar regularly during the computation. JAX-tqdm implements this for JAX scans and loops and is used by simply adding a decorator to the body of your update function.
```
Comment (jeremiecoullon) [inline README.md:71]: Unrelated to your PR, but what I had written here is slightly weird. Could you replace `JAX functions are purely functional` with `JAX functions are pure` please? (keeping the URL)
It should either be "JAX is purely functional", or "JAX functions are pure"; wrote I wrote is a weird halfway house :p

**Comment by jeremiecoullon** [comment]
Thinking about `print_rate` just now (I hadn't thought about this before). We talk in the readme about how the progress bar updates 20 times by default, but that the user can now override that.

Would it not rather make sense for the optional argument to be `num_updates` (which defaults to 20) rather than `print_rate`? 
However, if you've found it more natural to change the print rate directly in your use cases then let's keep this!

**Comment by zombie-einstein** [comment]
> Thinking about `print_rate` just now (I hadn't thought about this before). We talk in the readme about how the progress bar updates 20 times by default, but that the user can now override that.
> 
> Would it not rather make sense for the optional argument to be `num_updates` (which defaults to 20) rather than `print_rate`? However, if you've found it more natural to change the print rate directly in your use cases then let's keep this!

Yeah see your point. In my specific case I have an outer loop that takes several seconds to run, so I was just going to print every step. I figured `print_rate` saves the small bit of mental maths to work out `num_updates` for the specific case?

**Comment by jeremiecoullon** [comment]
Ok makes sense! Go for it.

On a seperate note, I've found that for very expensive `step` functions it's sometimes not worth running `scan`/`fori_loop` (and I think at times I've even found that it's slower than a python loop for some reason; but I haven't tested it rigorously). In those cases running a python loops works (as has much faster compilation speed). But if it works well for your example then there's no problem of course!

File: README.md [hunk 4]
```diff
@@ -64,13 +64,20 @@ def step(carry, x):
 last_number, all_numbers = lax.scan(step, 0, jnp.arange(n))
 ```
 
-will update every-other steps.
+will update every other step.
 
 ## Why JAX-tqdm?
 
-JAX functions are [purely functional](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#pure-functions), so side effects such as printing progress when running scans and loops are not allowed. However, the [host_callback module](https://jax.readthedocs.io/en/latest/jax.experimental.host_callback.html) has primitives for calling Python functions on the host from JAX code. This can be used to update a Python tqdm progress bar regularly during the computation. JAX-tqdm implements this for JAX scans and loops and is used by simply adding a decorator to the body of your update function.
```
Comment (jeremiecoullon) [inline README.md:48]: moar nitpick (sorry :) ): _JAX functions are pure_ (not purely)

Comment (jeremiecoullon) [reply README.md:48]: (I mean to add this comment in the edited code by the way)

Comment (zombie-einstein) [reply README.md:48]: Jeez sorry, duh! Will fix!

Comment (zombie-einstein) [reply README.md:48]: Fixed, slow brain on a Monday morning!

**Comment by zombie-einstein** [comment]
> Ok makes sense! Go for it.
> 
> On a seperate note, I've found that for very expensive `step` functions it's sometimes not worth running `scan`/`fori_loop` (and I think at times I've even found that it's slower than a python loop for some reason; but I haven't tested it rigorously). In those cases running a python loops works (as has much faster compilation speed). But if it works well for your example then there's no problem of course!

This is interesting, I was under the impression the opposite could be a pitfall, where JAX is compiling inside a loop for example. Will have to look into.

**Comment by jeremiecoullon** [comment]
>where JAX is compiling inside a loop for example.

Do you mean: in cases where JAX would recompile the `step` function at every iteration? That's not good for performance yeah; but if the input type/shapes don't change then it only compile the first time it's called.


But in any case, this PR looks good; feel free to merge!

**Pull request merged**